### PR TITLE
feat: add custom_api_max_price_age parameter

### PR DESCRIPTION
**Pull request by nvtaveras**
**BEFORE REVIEW**
Please note that this PR is pointing to be merged into master but this is only so that the CI runs on every new commit. We should instead merge this into the [v1.17.0 hotfix branch instead](https://github.com/mento-protocol/hummingbot/tree/v1.17.0-hotfix) (to not diverge master from upstream).

**TL;DR: Do not merge to master!**

--------
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

This PR adds a new `custom_api_max_price_age` which allows users to set a threshold for how old the price returned by a custom api can be before invalidating it. 

Before this change, the `CustomAPIDataFeed` would return a cached version of the price whenever there is any error when trying to fetch the price from the API. This is obviously not safe as the price could be outdated but still being used as the reference price for placing orders by the strategies (for example in the PMM). With the new parameter the price will essentially "expire" after certain time if it's not being updated successfully, and just return `NaN`. At the same time the PMM strategy will check that the price is not `NaN` before running the strategy steps (proposing/modifying orders, etc) and will just cancel any active orders for safety instead.

By default the new parameter will be disabled (-1) for backwards compatibility and to not introduce changes to any of the existing strategies.

**Tests performed by the developer**:

1. Wrote unit tests for the `CustomAPIDataFeed` which didn't exist before
2. Ran this locally using this simple python API that exposes methods for testing the behaviour of the bot in different circumstances:

`GET /price` get the latest price
`GET /set?price=0.5`set the latest price to 0.5
`GET /off` switch off /price for it to return 500 with a `NaN` response (to simulate failures)
`GET /on` switch on the /price endpoint to start returning 200 responses again and the price value

```python
from flask import Flask, request

app = Flask(__name__)

PRICE = 1.0
PRICE_STATE = True

@app.route('/price', methods=['GET'])
def get_price():
    if PRICE_STATE:
        return str(PRICE), 200
    else:
        return "NaN", 500

@app.route('/set', methods=['GET'])
def set_price():
    global PRICE
    PRICE = float(request.args.get('price', '1.0'))
    return "Price set", 200

@app.route('/off', methods=['GET'])
def off_price():
    global PRICE_STATE
    PRICE_STATE = False
    return "Switched off", 200

@app.route('/on', methods=['GET'])
def on_price():
    global PRICE_STATE
    PRICE_STATE = True
    return "Switched on", 200


if __name__ == '__main__':
    app.run(debug=True)
```

**Tips for QA testing**:

Run the PMM strategy pointing to the local API and try different scenarios like modifying the set price and flipping the endpoint on and off.

### CODE REVIEW

**Review by ninabarbakadze — APPROVED** [review]
looks good 👍🏼

**Comment by bowd** [comment]
I know this is non-standard but given we want to submit this PR upstream, I'm leaving some comments on the issue description as well. I think the first sentence after "A description of the changes proposed in the pull request:" is hard to read. I would rewrite those first few lines by starting with the status quo: there is this `CustomAPIDataFeed` which can feed a reference price into the strategies, if the API throws an error the class always returns the last valid response, this PR adds lifetime to the values collected from the API, and a setting to configure when a value should expire. 

I think starting with `custom_api_max_price_age` is jumping the gun a bit.

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx [hunk 1]
```diff
@@ -7,27 +7,37 @@ import numpy as np
...
+
 from hummingbot.strategy.order_book_asset_price_delegate cimport OrderBookAssetPriceDelegate
+
 from hummingbot.strategy.strategy_base import StrategyBase
 from hummingbot.strategy.utils import order_age
+
 from .data_types import PriceSize, Proposal
 from .inventory_cost_price_delegate import InventoryCostPriceDelegate
+
 from .inventory_skew_calculator cimport c_calculate_bid_ask_ratios_from_base_asset_ratio
```
Comment (bowd) [inline hummingbot/strategy/pure_market_making/pure_market_making.pyx:36]: What's up with all the white space here?

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx [hunk 2]
```diff
@@ -737,6 +747,16 @@ cdef class PureMarketMakingStrategy(StrategyBase):
                                           f"making may be dangerous when markets or networks are unstable.")
 
             proposal = None
+            ref_price_available = not self.get_price().is_nan()
+            if not ref_price_available:
```
Comment (bowd) [inline hummingbot/strategy/pure_market_making/pure_market_making.pyx]:
```suggestion
            ref_price_unavailable = self.get_price().is_nan()
            if ref_price_unavailable:
```

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx [hunk 3]
```diff
@@ -737,6 +747,16 @@ cdef class PureMarketMakingStrategy(StrategyBase):
...
+                self.logger().info("Reference price not available. Waiting for a new price to resume strategy...")
+
+                if len(self.active_orders) > 0:
+                    self.logger().info("Cancelling any active orders for safety...")
+                    self.c_cancel_all_active_orders()
+
+                return
```
Comment (bowd) [inline hummingbot/strategy/pure_market_making/pure_market_making.pyx]: there's this statement at the end of this function:
```python
        finally:
            self._last_timestamp = timestamp
```
That won't get executed if we short-circuit the return here. Is this a problem?

File: hummingbot/strategy/pure_market_making/pure_market_making_config_map.py [hunk 4]
```diff
@@ -379,6 +389,14 @@ def validate_decimal_list(value: str) -> Optional[str]:
                   default=float(5),
                   type_str="float",
                   validator=lambda v: validate_decimal(v, Decimal("0.5"))),
+    "custom_api_max_price_age":
+        ConfigVar(key="custom_api_max_price_age",
+                  prompt="Enter a time in seconds to invalidate the price from custom API if stale "
+                         "(Enter -1 to deactivate this feature) >>> ",
```
Comment (bowd) [inline hummingbot/strategy/pure_market_making/pure_market_making_config_map.py]:
```suggestion
                  prompt="Enter the number of seconds after which a price from the custom API becomes stale"
                                 "(default: -1, prices never expire) >>>",

```

File: hummingbot/strategy/pure_market_making/pure_market_making_config_map.py [hunk 5]
```diff
@@ -79,6 +79,16 @@ def validate_price_floor_ceiling(value: str) -> Optional[str]:
         return "Value must be more than 0 or -1 to disable this feature."
 
 
+def validate_custom_api_max_price_age(value: int) -> Optional[str]:
```
Comment (bowd) [inline hummingbot/strategy/pure_market_making/pure_market_making_config_map.py:82]: Looks like you can get related fields and based validation on that, and my understanding here is that
`value  >= pure_market_making_config_map.get("custom_api_update_interval").value` should be added as a condition.

**Review by bowd — CHANGES_REQUESTED** [review]
Looks good overall, left a few notes 📔

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx (hunk 1 above)
Comment (nvtaveras) [reply hummingbot/strategy/pure_market_making/pure_market_making.pyx:36]: They were added by the pre-commit hooks

File: hummingbot/strategy/pure_market_making/pure_market_making.pyx (hunk 3 above)
Comment (nvtaveras) [reply hummingbot/strategy/pure_market_making/pure_market_making.pyx]: good catch.. this seems to be used for checking if the logger should report warnings based on some interval:
https://github.com/mento-protocol/hummingbot/pull/1/files#diff-ff54f4274a270c937744102bfad74438719543f0f63cf3a45622e62877cfbb30L718-L721

**Comment by nvtaveras** [comment]
> I know this is non-standard but given we want to submit this PR upstream, I'm leaving some comments on the issue description as well. I think the first sentence after "A description of the changes proposed in the pull request:" is hard to read. I would rewrite those first few lines by starting with the status quo: there is this `CustomAPIDataFeed` which can feed a reference price into the strategies, if the API throws an error the class always returns the last valid response, this PR adds lifetime to the values collected from the API, and a setting to configure when a value should expire.
> 
> I think starting with `custom_api_max_price_age` is jumping the gun a bit.

Yeah this sounds reasonable to me. I was planning to rewrite the PR description before I submit it to upstream and ask for internal feedback. I agree that the way we frame this is important (since this seems like a bug rather than a new feature)

**Review by bowd — APPROVED** [review]

**Pull request merged**